client/client-lib/src/mcsm.rs

```diff
@@ -0,0 +1,76 @@
+//! # Module Client State Machine
+//! Module clients have to keep track of inputs and outputs created in the past in case further
+//! action is required. Since the client can crash at any point intermediate results have to be
+//! persisted atomically. This module defines an interface and execution environment for such
+//! actions by modeling them as state machines.
+
+use std::time::SystemTime;
+
+use async_trait::async_trait;
+use fedimint_api::db::DatabaseTransaction;
+use fedimint_api::TransactionId;
+use fedimint_core::transaction::Input;
+
+use crate::query::QueryStrategy;
+
+#[async_trait]
+pub trait State {
+    /// Returns a list of events on which the state machine should be woken up
+    fn poll_events(&self) -> Vec<PollEventCondition>;
+
+    /// Logical operation to which the action belongs.
+    ///
+    /// This information can be used to route anyway correlated network traffic over the same Tor
+    /// circuit. It is typically set to the transaction id of the first transaction of the
+    /// operation.
+    fn operation_id(&self) -> TransactionId;
+
+    /// Try to drive the state machine forward
+    async fn poll(
+        self: Box<Self>,
+        side_effects_executor: &mut dyn DelayedSideEffectsExecutor,
+        event_condition: PollEventCondition,
+        event: PollEvent,
+    ) -> Result<Box<Self>, anyhow::Error>;
+}
+
+pub enum PollEventCondition {
```

Comment:
This all feels so much like implementing async executor that it makes me think - should we possibly make these state `Future`s and let them "poll themselves" by the executor?

Reply:
The client code could kind of ... `select` on all the "pending states" (ones that can change automatically), until they resolve themselves to ... "stable states" or something like this?

I don't have a full vision of it yet, but just intuition that it's possible.

Reply:
> that can generate these futures,

I was thinking it will be something that *is* a `Future`. So you can serialize and store it, or you can `.await` it :)

Reply:
After more consideration I think that we want to keep the "wait till we should process" and "actually do stuff" separate. That way we can await all the potential state transitions in parallel and only once one of them becomes ready we give it a mut handle on a DB transaction etc. and only process one state transition in parallel. This way we can avoid some DB conflicts and don't run into hypothetical synchronization issues with mutable references to shared state.

I'm still undecided if I want to leave the wake-up condition as an enum for testability or if it should become a `Box<dyn Fn(&SomeApiInterface) -> Box<dyn Future<Output = serde_json::Value>>>`.

Reply:
Do the state-machines even need db transactions? Could they just be `Future`s that resolve to a new state that the client is responsible for storing it? Or maybe the new state should have a `fn store(&self, &mut DBTransaction) -> Result<()>`? to just save itself, just like mentioned  - awaiting new state separate, saving it separate.





> I'm still undecided if I want to leave the wake-up condition as an enum for testability or if it should become a `Box<dyn Fn(&SomeApiInterface) -> Box<dyn Future<Output = serde_json::Value>>>`.

I don't understand this part.